File: slidingWindow/leetcode_121/solution.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        self.max_profit = 0
        l, r = 0, 1
        
        while r < len(prices):
            profit = prices[r] - prices[l]
            self.max_profit = max(profit, self.max_profit)

            if prices[l] > prices[r]:
                l = r
            r += 1
        return self.max_profit
        

File: slidingWindow/leetcode_121/solution.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    l, r = 0, 1
    max_profit = 0
    while r < prices.length
        profit = prices[r] - prices[l]
        max_profit = [profit, max_profit].max
        if prices[l] > prices[r]
            l = r
        end
        r += 1
    end
    return max_profit
end


test_cases = [
    {
        prices: [7,1,5,3,6,4],
        result: 5
    },
    {
        prices: [7,6,4,3,1],
        result: 0
    },
]

for test_case in test_cases
    result = max_profit(test_case[:prices])
    if result == test_case[:result]
        print("Pass!\n")
    else
        print("Not pass, result #{result}\n")
    end
end